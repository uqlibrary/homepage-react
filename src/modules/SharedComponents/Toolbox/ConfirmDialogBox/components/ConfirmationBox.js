import React from 'react';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogTitle from '@material-ui/core/DialogTitle';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogContent from '@material-ui/core/DialogContent';
import { makeStyles } from '@material-ui/core/styles';
import Grid from '@material-ui/core/Grid';
import Hidden from '@material-ui/core/Hidden';

export const useStyles = makeStyles(theme => ({
    alternateActionButtonClass: {
        color: theme.palette.white.main,
        backgroundColor: theme.palette.warning.main,
        '&:hover': {
            backgroundColor: theme.palette.warning.dark,
        },
    },
}));

export const ConfirmationBox = ({
    actionButtonColor,
    actionButtonVariant,
    cancelButtonColor,
    confirmationBoxId,
    InputForm,
    hideActionButton,
    hideCancelButton,
    isOpen,
    locale,
    onAction,
    onAlternateAction,
    onCancelAction,
    onClose,
    showAlternateActionButton,
    showInputForm,
}) => {
    const classes = useStyles();

    const _onAction = () => {
        onClose();
        onAction();
    };

    const _onCancelAction = () => {
        onClose();
        !!onCancelAction && onCancelAction();
    };

    const _onAlternateAction = () => {
        onClose();
        !!onAlternateAction && onAlternateAction();
    };

    return (
        <Dialog style={{ padding: 6 }} open={isOpen} data-testid={`dialogbox-${confirmationBoxId}`}>
            <DialogTitle data-testid="message-title">{locale.confirmationTitle}</DialogTitle>
            <DialogContent style={{ minWidth: 400 }}>
                <DialogContentText data-testid="message-content">{locale.confirmationMessage}</DialogContentText>
                {!!showInputForm && /* istanbul ignore next */ <InputForm />}
            </DialogContent>
            <DialogActions>
                <Grid container spacing={1}>
                    <Hidden xsDown>
                        <Grid item xs />
                    </Hidden>
                    {!hideActionButton && (
                        <Grid item xs={12} sm={'auto'}>
                            <Button
                                {...(!!actionButtonVariant ? { variant: actionButtonVariant } : {})}
                                children={locale.confirmButtonLabel}
                                autoFocus
                                color={actionButtonColor || 'primary'}
                                fullWidth
                                onClick={_onAction}
                                id="confirm-action"
                                data-testid="confirm-action"
                            />
                        </Grid>
                    )}
                    {showAlternateActionButton && (
                        // an optional middle button that will display in a warning colour
                        <Grid item xs={12} sm={'auto'}>
                            <Button
                                variant={'contained'}
                                className={classes.alternateActionButtonClass}
                                children={locale.alternateActionButtonLabel}
                                fullWidth
                                onClick={_onAlternateAction}
                                id="confirm-alternate-action"
                                data-testid="confirm-alternate-action"
                            />
                        </Grid>
                    )}
                    {!hideCancelButton && (
                        <Grid item xs={12} sm={'auto'}>
                            <Button
                                variant={'contained'}
                                color={cancelButtonColor || 'primary'}
                                children={locale.cancelButtonLabel}
                                fullWidth
                                onClick={_onCancelAction}
                                id="confirm-cancel-action"
                                data-testid="confirm-cancel-action"
                            />
                        </Grid>
                    )}
                </Grid>
            </DialogActions>
        </Dialog>
    );
};

ConfirmationBox.propTypes = {
    actionButtonColor: PropTypes.string,
    actionButtonVariant: PropTypes.string,
    cancelButtonColor: PropTypes.string,
    confirmationBoxId: PropTypes.string.isRequired,
    hideActionButton: PropTypes.bool,
    hideCancelButton: PropTypes.bool,
    InputForm: PropTypes.func,
    isOpen: PropTypes.bool,
    locale: PropTypes.object,
    onAction: PropTypes.func,
    onCancelAction: PropTypes.func,
    onAlternateAction: PropTypes.func,
    onClose: PropTypes.func,
    showAlternateActionButton: PropTypes.bool,
    showInputForm: PropTypes.bool,
};

ConfirmationBox.defaultProps = {
    hideActionButton: false,
    hideCancelButton: false,
    isOpen: false,
    locale: {
        confirmationTitle: 'Confirmation',
        confirmationMessage: 'Are you sure?',
        cancelButtonLabel: 'No',
        confirmButtonLabel: 'Yes',
        alternateActionButtonLabel: 'Cancel',
    },
    showAlternateActionButton: false,
    showInputForm: false,
};

export default React.memo(ConfirmationBox);
