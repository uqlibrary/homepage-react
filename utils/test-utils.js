/* eslint-disable react/prop-types */
import React from 'react';
import PropTypes from 'prop-types';
import { render } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { Route } from 'react-router';
import { mui1theme } from 'config/theme';
import { Provider } from 'react-redux';
import { ThemeProvider as MuiThemeProvider } from '@material-ui/core/styles';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import MomentUtils from '@date-io/moment';

import preview from 'jest-preview';

import { getStore } from '../src/config/store';
import Immutable from 'immutable';
import { createMemoryHistory } from 'history';

const domTestingLib = require('@testing-library/dom');
const reactTestingLib = require('@testing-library/react');

const { configure } = domTestingLib;

configure(config => ({
    ...config,
    testIdAttribute: 'id',
}));

const AllTheProviders = props => {
    return (
        <MuiThemeProvider theme={mui1theme}>
            <MuiPickersUtilsProvider utils={MomentUtils}>{props.children}</MuiPickersUtilsProvider>
        </MuiThemeProvider>
    );
};

AllTheProviders.propTypes = {
    children: PropTypes.node,
};

export const rtlRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options });

export const renderWithRouter = (
    ui,
    { route = '/', history = createMemoryHistory({ initialEntries: [route] }) } = {},
) => {
    return render(
        <AllTheProviders>
            <Router history={history}>{ui}</Router>
        </AllTheProviders>,
    );
};

export const RenderWithRouter = ({
    children,
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
}) => (
    <AllTheProviders>
        <Router history={history}>{children}</Router>
    </AllTheProviders>
);
export const renderWithRedux = ({ initialState }) => render => {
    return {
        ...render,
        store: getStore({ initialState, history: render.history }),
    };
};

export const withRouter = ({
    route = '/',
    path = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
} = {}) => WrappedComponent => {
    return (
        <Router history={history}>
            <Route path={path} children={WrappedComponent} />
        </Router>
    );
};

export const withRedux = (initialState = Immutable.Map()) => WrappedComponent => {
    return <Provider store={getStore(initialState)}>{WrappedComponent}</Provider>;
};

// eslint-disable-next-line react/prop-types
export const WithRedux = ({ initialState = Immutable.Map(), children }) => (
    <Provider store={getStore(initialState)}>{children}</Provider>
);

export const WithReduxStore = ({ initialState = Immutable.Map(), children }) => (
    <Provider store={getStore(initialState)}>
        <AllTheProviders>{children}</AllTheProviders>
    </Provider>
);

module.exports = {
    ...domTestingLib,
    ...reactTestingLib,
    rtlRender,
    renderWithRouter,
    renderWithRedux,
    withRedux,
    withRouter,
    WithRedux,
    WithReduxStore,
    RenderWithRouter,
    preview,
};
