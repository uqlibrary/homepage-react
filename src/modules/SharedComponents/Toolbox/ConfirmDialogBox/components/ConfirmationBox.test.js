import React from 'react';
import ConfirmationBox from './ConfirmationBox';
import { rtlRender, fireEvent, act } from 'test-utils';

function setup(testProps = {}, renderer = rtlRender) {
    const props = {
        hideCancelButton: false,
        isOpen: false,
        locale: {
            confirmationTitle: 'Confirmation',
            confirmationMessage: 'Are you sure?',
            cancelButtonLabel: 'No',
            confirmButtonLabel: 'Yes',
            alternateActionButtonLabel: 'Cancel',
        },
        onAction: jest.fn(),
        onAlternateAction: jest.fn(),
        onCancelAction: jest.fn(),
        onClose: jest.fn(),
        showAlternateActionButton: false,
        confirmationBoxId: 'confirmation-box',
        ...testProps,
    };

    return renderer(<ConfirmationBox {...props} />);
}

describe('ConfirmationBox component', () => {
    it('should render confirmation box', () => {
        const { getByTestId, getByText } = setup({ isOpen: true });
        expect(getByText('Are you sure?')).toBeInTheDocument();
        expect(getByTestId('confirm-action')).toBeInTheDocument();
        expect(getByTestId('confirm-cancel-action')).toBeInTheDocument();
    });

    it('should render alternate action button', () => {
        const { getByTestId, getByText } = setup({ isOpen: true, showAlternateActionButton: true });
        expect(getByText('Are you sure?')).toBeInTheDocument();
        expect(getByTestId('confirm-action')).toBeInTheDocument();
        expect(getByTestId('confirm-cancel-action')).toBeInTheDocument();
        expect(getByTestId('confirm-alternate-action')).toBeInTheDocument();
    });

    it('should call confirm action', () => {
        const onActionFn = jest.fn();
        const onCloseFn = jest.fn();
        const { getByTestId } = setup({ isOpen: true, onAction: onActionFn, onClose: onCloseFn });
        fireEvent.click(getByTestId('confirm-action'));
        expect(onActionFn).toBeCalled();
        expect(onCloseFn).toBeCalled();
    });

    it('should call cancel action', () => {
        const onCancelActionFn = jest.fn();
        const onCloseFn = jest.fn();
        const { getByTestId } = setup({ isOpen: true, onCancelAction: onCancelActionFn, onClose: onCloseFn });
        fireEvent.click(getByTestId('confirm-cancel-action'));
        expect(onCancelActionFn).toBeCalled();
        expect(onCloseFn).toBeCalled();
    });

    it('should call alternate action', () => {
        const onAlternateActionFn = jest.fn();
        const onCloseFn = jest.fn();
        const { getByTestId } = setup({
            isOpen: true,
            onAlternateAction: onAlternateActionFn,
            onClose: onCloseFn,
            showAlternateActionButton: true,
        });
        fireEvent.click(getByTestId('confirm-alternate-action'));
        expect(onAlternateActionFn).toBeCalled();
        expect(onCloseFn).toBeCalled();
    });
});
