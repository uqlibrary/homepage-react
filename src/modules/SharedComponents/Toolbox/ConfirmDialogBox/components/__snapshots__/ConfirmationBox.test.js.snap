// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmationBox component default values render correctly 1`] = `<div />`;
